lib/evaluate/rte.cc

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2018-2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file defines the runtime libraries for the target as well as a default
+// set of host rte functions that can be used for folding.
+// The default HostRte is built with <cmath> and <complex> functions
+// that are guaranteed to exist from the C++ standard.
+
+#include "rte.h"
+#include <cerrno>
+#include <cfenv>
+
+namespace Fortran::evaluate::rte {
+
+using namespace Fortran::parser::literals;
+// TODO mapping to <cmath> function to be tested.<cmath> func takes
+// real arg for n
+template<typename HostT> static HostT Bessel_jn(std::int64_t n, HostT x) {
+  return std::cyl_bessel_j(static_cast<HostT>(n), x);
+}
+template<typename HostT> static HostT Bessel_yn(std::int64_t n, HostT x) {
+  return std::cyl_neumann(static_cast<HostT>(n), x);
+}
+
+template<typename HostT> void AddLibmRealHostProcedure(HostRte &hostRte) {
+  using F = FuncPointer<HostT, HostT>;
+  using F2 = FuncPointer<HostT, HostT, HostT>;
+  HostRteProcedureSymbol libmSymbols[]{{"acos", F{std::acos}, true},
```

Comment:
Should there be an enum of intrinsics so we don't have to do string comparisons on the names?

Reply:
If we reduce the number of hand written comparisons on the names by better delegating it to a map lookup when there is no need to do something particular, then I am not sure it is worth having to define such an enum somewhere and having to map it with the string names and maintain it. I will try to go with the hostRte.TryFolding(funcRef) proposal that should greatly reduce the number of hand written string comparison first.